File: source/math/Ray.java
package math;

import java.io.Serializable;

public class Ray implements Serializable {
    public Float3 origin;
    public Float3 direction;

    public Ray(Float3 origin, Float3 direction) {
        this.origin = origin;
        this.direction = direction;
    }

    public Ray(Ray ray) {
        origin = new Float3(ray.origin);
        direction = new Float3(ray.direction);
    }

    public boolean intersect(Plane plane, Float3 outIntersection) {
        final float denom = plane.normal.normalize().dot(direction.normalize());

        if (Math.abs(denom) > 0.0001f) {
		    final float t = plane.point.subtract(origin).dot(plane.normal) / denom;

            if (t >= 0.0f) {
                if (outIntersection != null) {
                    outIntersection.x = origin.x + direction.x * t;
                    outIntersection.y = origin.y + direction.y * t;
                    outIntersection.z = origin.z + direction.z * t;
                }
                return true;
            }
        }

        return false;
    }

    public boolean intersect(Triangle triangle, Float3 outIntersection) {
        final Float3 edge1 = triangle.b.world.subtract(triangle.a.world);
        final Float3 edge2 = triangle.c.world.subtract(triangle.a.world);
    
        final Float3 h = direction.cross(edge2);
        final Float3 s = origin.subtract(triangle.a.world);
        final float f = 1 / edge1.dot(h);
        final float u = s.dot(h) * f;
        if (u < 0 || u > 1) {
            return false;
        }

        final Float3 q = s.cross(edge1);
        final float v = direction.dot(q) * f;
        if (v < 0 || (u + v) > 1) {
            return false;
        }

        final float t = edge2.dot(q) * f;
        if (t > 0) {
            if (outIntersection != null) {
			    outIntersection.x = origin.x + direction.x * t;
                outIntersection.y = origin.y + direction.y * t;
                outIntersection.z = origin.z + direction.z * t;
            }
            return true;
        }
        return false;
    }

    public boolean intersectSphere(Sphere sphere, Float4 outData) {
        final Float3 centerRay = sphere.center.subtract(origin);
        final float cdDot = centerRay.dot(direction);
        if (cdDot < 0) {
            return false;
        }

        final float ccDot = centerRay.dot(centerRay) - cdDot * cdDot;
        final float rr = sphere.radius * sphere.radius;
        if (ccDot > rr) {
            return false;
        }

        final float thc = (float) Math.sqrt(rr - ccDot);
        final float dis0 = cdDot - thc;
        final float dis1 = cdDot + thc;

        outData.w = ((dis0 < 0) ? dis1 : dis0);
        outData.x = origin.x + direction.x * outData.w;
        outData.y = origin.y + direction.y * outData.w;
        outData.z = origin.z + direction.z * outData.w;

        return true;
    }

    public boolean intersectSphere(Sphere sphere) {
        final float rayLength = sphere.center.subtract(origin).dot(direction);
        final Float3 rayPoint = direction.multiply(rayLength).add(origin);
        final float sphereRayDistance = sphere.center.subtract(rayPoint).length();
        return !(sphereRayDistance > sphere.radius);
    }
}

File: source/math/Ray.kt
package math

import java.io.Serializable
import kotlin.math.abs
import kotlin.math.sqrt

class Ray : Serializable {
    var origin: Float3 = Float3()
    var direction: Float3 = Float3()

    constructor()

    constructor(origin: Float3, direction: Float3) {
        this.origin = origin
        this.direction = direction
    }

    constructor(ray: Ray) {
        origin = Float3(ray.origin)
        direction = Float3(ray.direction)
    }

    fun intersect(plane: Plane, outIntersection: Float3): Boolean {
        val denom = normalize(plane.normal) * normalize(direction)
        if (abs(denom) > 0.0001f) {
            val t = (plane.point - origin) * plane.normal / denom
            if (t >= 0f) {
                val res = origin + direction * t
                outIntersection.x = res.x
                outIntersection.y = res.y
                outIntersection.z = res.z
                return true
            }
        }
        return false
    }

    fun intersect(triangle: Triangle, outIntersection: Float3): Boolean {
        val edge1 = triangle.b.world - triangle.a.world
        val edge2 = triangle.c.world - triangle.a.world

        val h = direction x edge2
        val s = origin - triangle.a.world
        val f = 1f / (edge1 * h)
        val u = s * h * f
        if (u < 0f || u > 1f) {
            return false
        }

        val q = s x edge1
        val v = direction * q * f
        if (v < 0f || (u + v) > 1f) {
            return false
        }

        val t = edge2 * q * f
        if (t > 0f) {
            val res = origin + direction * t
            outIntersection.x = res.x
            outIntersection.y = res.y
            outIntersection.z = res.z
            return true
        }
        return false
    }

    fun intersectSphere(sphere: Sphere, outData: Float4? = null): Boolean {
        val centerRay = sphere.center - origin
        val cdDot = centerRay * direction
        if (cdDot < 0f) {
            return false
        }

        val ccDot = centerRay * centerRay - cdDot * cdDot
        val rr = sphere.radius * sphere.radius
        if (ccDot > rr) {
            return false
        }

        val thc = sqrt(rr - ccDot)
        val dis0 = cdDot - thc
        val dis1 = cdDot + thc
        outData?.let {
            it.w = if (dis0 < 0f) dis1 else dis0
            it.x = origin.x + direction.x * it.w
            it.y = origin.y + direction.y * it.w
            it.z = origin.z + direction.z * it.w
        }
        return true
    }

    fun intersectSphere(sphere: Sphere): Boolean {
        val rayLength = (sphere.center - origin) * direction
        val rayPoint = direction * rayLength + origin
        val sphereRayDistance = (sphere.center - rayPoint).length
        return sphereRayDistance <= sphere.radius
    }
}